app/models/criterion.rb

```diff
@@ -1,46 +1,65 @@
 # The abstract base class that defines common behavior for all types of
 # criterion.
 class Criterion < ApplicationRecord
+  belongs_to :assignment, foreign_key: :assessment_id
   after_update :scale_marks
 
+  has_many :marks, dependent: :destroy
+  accepts_nested_attributes_for :marks
+
+  validates_presence_of :assigned_groups_count
+  validates_numericality_of :assigned_groups_count
+  before_validation :update_assigned_groups_count
+
+
+  has_many :criterion_ta_associations, dependent: :destroy
+  has_many :tas, through: :criterion_ta_associations
+  has_many :test_groups
+
   validates_presence_of :name
   validates_uniqueness_of :name, scope: :assessment_id
 
   validates_presence_of :max_mark
   validates_numericality_of :max_mark, greater_than: 0
 
   has_many :criteria_assignment_files_joins,
-           as: :criterion,
            dependent: :destroy
+
   has_many :assignment_files,
            through: :criteria_assignment_files_joins
   accepts_nested_attributes_for :criteria_assignment_files_joins, allow_destroy: true
 
   validate :results_unreleased?
   validate :visible?
 
-  self.abstract_class = true
+  has_many :levels, -> { order(:mark) }, inverse_of: :criterion, dependent: :destroy, autosave: true
```

Comment:
Hmm, this association still belongs in the `RubricCriterion` model, I think. I can't remember if you ran into an issue keeping it in `RubricCriterion`---if so, let me know, and if not, move it back there.

(Similar with line below)

Reply:
I'll keep you updated when I start doing UI testing!